ml-agents/mlagents/trainers/trainer_controller.py

```diff
@@ -233,40 +231,55 @@ def start_learning(
         for _, t in self.trainers.items():
             self.logger.info(t)
 
+        global_step = 0
+
         if self.train_model:
             for brain_name, trainer in self.trainers.items():
                 trainer.write_tensorboard_text("Hyperparameters", trainer.parameters)
         try:
-            curr_info = self._reset_env(env)
+            for brain_name, trainer in self.trainers.items():
+                env_manager.set_policy(brain_name, trainer.policy)
+            self._reset_env(env_manager)
             while (
                 any([t.get_step <= t.get_max_steps for k, t in self.trainers.items()])
                 or not self.train_model
             ):
-                new_info = self.take_step(env, curr_info)
-                self.global_step += 1
-                if (
-                    self.global_step % self.save_freq == 0
-                    and self.global_step != 0
-                    and self.train_model
-                ):
-                    # Save Tensorflow model
-                    self._save_model(steps=self.global_step)
-                curr_info = new_info
+                n_steps = self.take_step(env_manager)
+                for i in range(n_steps):
+                    global_step += 1
+                    if (
```

Comment:
Maybe encapsulate this as `_should_save_model()`?

Reply:
😁 I actually have a branch where I did that already.  I don't see a problem in fitting it into this PR